Node examples (#7)
js-choi/proposal-bind-this: README.md

(A [formal specification][] is available.)

**Method binding** `->` is a **left-associative infix operator**.

> Its right-hand side is a **chain of identifiers identifier** (like `f` or `o.x.y`)

Reviewer: ```suggestion
Its right-hand side is a **chain of identifiers**, **identifier** (like `f` or `o.x.y`),
```

> or a parenthesized **expression** (like `hof()`),

Reviewer: this is a function invocation, not a parenthesized expression (that would be an expression wrapped in parens)

Author: Ah, of course. I’ll wrap `hof()` in parentheses `(hof())`.

Author: Fixed in e8537951f21a8f2b3575c999003f3568018760d7.

> or an **expression** in `(` `)` (like `hof()`),

Reviewer: the example is a function invocation, not an expression in parens

Author: Ah, this is an inconsistency. It should just be “identifier”; no `o.x.y` allowed.

I had been planning to allow method chains on the RHS (see https://github.com/js-choi/proposal-bind-this-operator/issues/3#issuecomment-920563136), but I decided against this for now to keep precedence simple. I am open to properly changing the RHS to allow property chains later.

**Method binding** `->` is a **left-associative infix operator**.
Its right-hand side is an **identifier** (like `f`)

> or a parenthesized **expression** (like `(hof())`),

Reviewer: this suggests i can't do `a->f()`, or are you saying that the `a->f` gets resolved before the function call?

Author: Yes, `a->f()` is equivalent to `(a->f)()`, not `a->(f())`.

`(a->f)()` is equivalent to `f.bind(a)()`, which is equivalent to `f.call(a)`.

If the operator’s right-hand side does not evaluate to a function during runtime,
then the program throws a `TypeError`.

The bound functions created by the bind-`this` operator
are **indistinguishable** from the bound functions
that are already created by [`Function.prototype.bind`][bind].
Both are **exotic objects** that do not have a `prototype` property,
and which may be called like any typical function.

From this definition, `o->f(...args)`

> is **indistinguishable** from `Function.prototype.call(o, ...args)`,

Reviewer 2: ```suggestion
is **indistinguishable** from `Function.prototype.call(f, o, ...args)`,
```

Reviewer: technically this isn't accurate either - you need `f.call(o, ...args)`, because if you just call `Function.prototype.call` directly then its receiver is `Function.prototype`.

> is **indistinguishable** from `Function.prototype.call(f, o, ...args)`,

Reviewer: ```suggestion
is **indistinguishable** from `f.call(o, ...args)`,
```
the current code has `Function.prototype` as the receiver for `.call`.

// Compare with extensionMethod.call(obj).
```

## Real-world examples
Only minor formatting changes have been made to the status-quo examples.

> ### Node.js

Author: Thanks; this is a good point. I think I’ll add a separate subsection about Deno with another example.

that are vulnerable to mutation or prototype pollution by third-party libraries.
When initializing a JavaScript runtime, Node.js therefore caches
wrapped versions of all global intrinsic objects (and their methods)
in a [large `primordials` object][primordials.js].
This `primordials` object contains numerous entries that look like this:
```js
ArrayPrototypeConcat: uncurryThis(Array.prototype.concat),
ArrayPrototypeCopyWithin: uncurryThis(Array.prototype.copyWithin),
ArrayPrototypeFill: uncurryThis(Array.prototype.fill),
ArrayPrototypeFind: uncurryThis(Array.prototype.find),
ArrayPrototypeFindIndex: uncurryThis(Array.prototype.findIndex),
ArrayPrototypeLastIndexOf: uncurryThis(Array.prototype.lastIndexOf),
ArrayPrototypePop: uncurryThis(Array.prototype.pop),
ArrayPrototypePush: uncurryThis(Array.prototype.push),
ArrayPrototypePushApply: UncurryThisStaticApply<typeof Array.prototype.push>
ArrayPrototypeReverse: uncurryThis(Array.prototype.reverse),
```
…and so on, where `uncurryThis` is `Function.bind.bind(Function.call)`
(also called [“call-binding”][call-bind]).

[call-bind]: https://npmjs.com/call-bind

> In other words, Node.js must **wrap** every global intrinsic method

Reviewer 2: ```suggestion
In other words, Node.js must **wrap** every `this` sensitive intrinsic method
```

some of the things we wrap are not globals and we don't wrap every method, some static methods are fine

[call-bind]: https://npmjs.com/call-bind

In other words, Node.js must **wrap** every `this`-sensitive global intrinsic method
in a `this`-uncurried **wrapper function**,
whose first argument is the method’s `this` value,
using the `uncurryThis` helper function.

The result is that code that uses these global intrinsic methods,
like this code adapted from [node/lib/internal/v8_prof_processor.js][]:
```js
  // specifier is a string.
  const file = specifier.slice(2, -4);
```
…
```js
  if (process.platform === 'darwin') {
    tickArguments.push('--mac');
  } else if (process.platform === 'win32') {
    tickArguments.push('--windows');
  }
  tickArguments.push(...process.argv.slice(1));
```
…must instead look like this:
```js
// Note: This module assumes that it runs before any third-party code.
const {
  ArrayPrototypePush,
  ArrayPrototypeSlice,
  StringPrototypeSlice,
} = primordials;
```
…
```js
  const file = StringPrototypeSlice(specifier, 2, -4);
```
…
```js
  if (process.platform === 'darwin') {
    ArrayPrototypePush(tickArguments, '--mac');
  } else if (process.platform === 'win32') {
    ArrayPrototypePush(tickArguments, '--windows');
  }

>   ArrayPrototypePush(tickArguments,
>                      ...ArrayPrototypeSlice(process.argv, 1));

Reviewer 3: We would use `ArrayPrototypePushApply` to avoid the unsafe array iteration here.
```suggestion
  ArrayPrototypePushApply(tickArguments,
                     ArrayPrototypeSlice(process.argv, 1));
```

Author: These lines were copied directly from <https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/v8_prof_processor.js>, which is up to date. Is the plan to change the code in that file, too?

Also, would the unsafe-array-iteration problem also apply to `tickArguments->$push(...process.argv->$arraySlice(1))`?

Reviewer: `...` can't be safely used, so the `pushApply` pattern would probably still need a custom bind.

Yes, it would still apply, because `...` invokes a `Symbol.iterator` method on the operand.

Author: Ah, I see. So that would require a wrapper function anyway.

I’ll probably just replace this example with node/lib/internal/debugger/inspect.js. I’ll push a new commit in a bit.

Reviewer: yes, that would work great

Author: Fixed in 04e752738c7ca80d5bae018f5dac94112530976a.

Reviewer 3: > Is the plan to change the code in that file, too? I’d rather have the example match how Node is written right now, but I’m fine with this change too.

Addressed in https://github.com/nodejs/node/pull/40271.

```

## Real-world examples
Only minor formatting changes have been made to the status-quo examples.

### Node.js
Node.js’s runtime depends on many built-in JavaScript global intrinsic objects
that are vulnerable to mutation or prototype pollution by third-party libraries.
When initializing a JavaScript runtime, Node.js therefore caches
wrapped versions of all global intrinsic objects (and their methods)
in a [large `primordials` object][primordials.js].
This `primordials` object contains numerous entries that look like this:
```js
ArrayPrototypeConcat: uncurryThis(Array.prototype.concat),
ArrayPrototypeCopyWithin: uncurryThis(Array.prototype.copyWithin),
ArrayPrototypeFill: uncurryThis(Array.prototype.fill),
ArrayPrototypeFind: uncurryThis(Array.prototype.find),
ArrayPrototypeFindIndex: uncurryThis(Array.prototype.findIndex),
ArrayPrototypeLastIndexOf: uncurryThis(Array.prototype.lastIndexOf),
ArrayPrototypePop: uncurryThis(Array.prototype.pop),
ArrayPrototypePush: uncurryThis(Array.prototype.push),
ArrayPrototypePushApply: UncurryThisStaticApply<typeof Array.prototype.push>
ArrayPrototypeReverse: uncurryThis(Array.prototype.reverse),
```

> …and so on, where `uncurryThis` is `Function.bind.bind(Function.call)`.

Reviewer: ```suggestion
…and so on, where `uncurryThis` is `Function.bind.bind(Function.call)` (also called ["call-binding"](https://npmjs.com/call-bind)).
```

(uncurryThis is a name used in node's primordials, born from SES's usage of it, but is not a term I've ever heard used outside of those domains)

is not much wordier than\
`const slice = arr&.slice; slice(1, 3);`

**Extracting property accessors** (i.e., getters and setters)
is not a goal of this proposal.
Get/set accessors are **not like** methods. Methods are **values**.
Accessors themselves are **not values**;
they are functions that activate when getting or setting properties.
Getters/setters have to be extracted using `Object.getOwnPropertyDescriptor`;
they are not handled in a special way.
This verbosity may be considered to be desirable [syntactic salt][]:
it makes the developer’s intention (to extract getters/setters – and not methods)
more explicit.

```js
const { get: $getSize } =
  Object.getOwnPropertyDescriptor(
    Set.prototype, 'size');

// The adversary’s code.
delete Set; delete Function;

// Our own trusted code, running later.

> new Set([0, 1, 2])->$getSize();

Reviewer 3: FWIW this is also dependent on `Array.prototype[Symbol.iterator]` (and `%ArrayIteratorPrototype%.next`) but maybe it's ok for this example.

Author: Yeah, let’s leave it for now in this pull request. 
If you come up with a more robust example, I’d be happy to merge a separate pull request. Thank you again for reviewing. : )
